### PR TITLE
Fix 'fromCodePointInt() should return error for invalid codePoints' and 'Char' /'String' incompatibility in langlib calls on literals

### LINKED ISSUE #31777: Cast is necessary for string literals to call toCodePointInt()
Consider the following code

`final int CP_BACKSLASH = ("\\").toCodePointInt();`

This gives error `incompatible types: expected 'string:Char', found 'string'`

### LINKED ISSUE #31778: fromCodePointInt() should return error for invalid codePoints
As per the current implementation fromCodePoint() does not return an error for surrogates (0xD800 to 0xDFFF).

It should give an error for surrogates

### PR DESCRIPTION
**Pull request by gimantha**
Fixes #31778
Fixes #31777

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 1]
```diff
@@ -631,7 +631,7 @@ private BType setLiteralValueAndGetType(BLangLiteral literalExpr, BType expType)
         } else if (literalType.tag == TypeTags.DECIMAL) {
             return decimalLiteral(literalValue, literalExpr, expType);
         } else if (literalType.tag == TypeTags.STRING && types.isCharLiteralValue((String) literalValue)) {
-            if (expType.tag == TypeTags.CHAR_STRING) {
+            if (expType.tag == TypeTags.CHAR_STRING || expType == symTable.noType) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:634]: Shouldn't the following still work?

```ballerina
public function main() {
    var s = "a";
    s = "hello";
}
```

Seems to result in an error (`incompatible types: expected 'string:Char', found 'string'`) with these changes.

Related section in the spec - https://ballerina.io/ballerina-spec/spec.html#section_6.2.3

Comment (gimantha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: fixed

**Review by KRVPerera — APPROVED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 2]
```diff
@@ -6037,7 +6037,16 @@ private BType checkInvocationArgs(BLangInvocation iExpr, List<BType> paramTypes,
             // function call error.
             if (i == 0 && arg.typeChecked && iExpr.expr != null && iExpr.expr == arg) {
                 BType expectedType = paramTypes.get(i);
-                types.checkType(arg.pos, arg.getBType(), expectedType, DiagnosticErrorCode.INCOMPATIBLE_TYPES);
+                BType actualType = arg.getBType();
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:6040]: @KRVPerera @MaryamZi I hope this fix is fine, as this is very specific to string:Char

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 3]
```diff
@@ -6037,7 +6037,16 @@ private BType checkInvocationArgs(BLangInvocation iExpr, List<BType> paramTypes,
...
+                if (expectedType == symTable.charStringType) {
+                    arg.cloneAttempt++;
+                    BLangExpression clonedArg = nodeCloner.cloneNode(arg);
+                    BType argType = checkExprSilent(clonedArg, expectedType, env);
+                    if (argType != symTable.semanticError) {
+                        actualType = argType;
+                    }
+                }
+                types.checkType(arg.pos, actualType, expectedType, DiagnosticErrorCode.INCOMPATIBLE_TYPES);
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:6049]: @KRVPerera @MaryamZi I hope this fix is fine, as this is very specific to string:Char

**Pull request merged**